Add tempOutput option to Pipeline.runOne()

rodneykinney opened the pull request.

rodneykinney committed:
Setting version to 1.2.1

rodneykinney committed:
Setting version to 1.2.2-SNAPSHOT

rodneykinney committed:
Add tmpOutput option to runOne()

rodneykinney committed:
Merge branch 'master' of https://github.com/allenai/pipeline

cristipp commented on src/main/scala/org/allenai/pipeline/Pipeline.scala:
```diff
@@ -35,15 +33,35 @@ trait Pipeline extends Logging {
+    val targetWithOverriddenLocation: Producer[T] =
+      tempOutput match {
+        case Some(tmp) =>
+          if (target.isInstanceOf[PersistedProducer[T, _ <: FlatArtifact]]) {
```
Are you sure instanceOf can handle all these cases at runtime? 

I'm thinking maybe it is possible to have PersistedProducer.redirect() { new PP(step) } which just works regardless of the Artifact style?

Edit: Looking at the code, most likely it will be handy to have Artifact.clone(newPath: String) and have a specialized impl for each Artifact instance (FlatArtifact, StructuredArtifact)

rodneykinney replied:
Since it's not a container, the type is not erased.  The checks work correctly in the unit tests.

rodneykinney replied:
If PersistedProducer were a case class, we could copy it and override the artifact.  This would be simpler for sure.  I don't see any downside to changing it to a case class.

cristipp commented on src/test/scala/org/allenai/pipeline/TestPipeline.scala:
```diff
@@ -88,4 +74,23 @@ class TestPipeline extends UnitSpec with ScratchDirectory {
+    val p2 = pipeline.persist(AddOne(p1), format, Some("prod2"))
+
+    val tmpOutput = new File(outputDir, "temp-output/prod2")
+    pipeline.runOne(p2, Some(tmpOutput.toURI.toString))
```
This test case looks great :)

rodneykinney committed:
Change type of output override in runOne()

rodneykinney commented:
Changed runOne() to accept an Artifact instead of a String.

Thinking of ways to make this simpler still.

rodneykinney committed:
Change runOne() signature

rodneykinney committed:
Cleanup

rodneykinney commented:
I think getting rid of the `RddPersister` interface will simplify things, because this is what resolves relative paths, which is making it harder to specify absolute paths.

cristipp commented:
As long as it makes delivering the functionality sooner rather than later,
let's go with it. Can you please give the command we're going to use to use
this method for RDDs?

Thanks!

On Tue, May 12, 2015 at 5:26 PM, rodneykinney notifications@github.com
wrote:

> Changed runOne() to accept an Artifact instead of a String.
> 
> Thinking of ways to make this simpler still.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/allenai/pipeline/pull/26#issuecomment-101467064.

vha14 commented on src/main/scala/org/allenai/pipeline/Pipeline.scala:
```diff
@@ -35,15 +33,27 @@ trait Pipeline extends Logging {
+    runOnly(
+      targetWithOverriddenLocation.stepInfo.className,
+      List(targetWithOverriddenLocation): _*
+    ).toList(0).asInstanceOf[T]
```
`head` vs `toList(0)`

rodneykinney committed:
Refactor ArtifactFactory

rodneykinney committed:
Cleanup

rodneykinney committed:
Refactor ArtifactFactory

rodneykinney committed:
Refactory ArtifactFactory

rodneykinney committed:
Add S3Config factory method

rodneykinney committed:
ExecuteShellCommand

rodneykinney committed:
Refactor ExecuteShellCommand

rodneykinney committed:
Refactor ExecuteShellCommand

rodneykinney committed:
Merge branch 'master' into artifact-factory

rodneykinney committed:
Add implicit conversions, BasicPipeline example

rodneykinney committed:
Remove path parameter from persist().  Update README

rodneykinney committed:
Improve HTML

rodneykinney committed:
Reformat

rodneykinney committed:
ExecutionInfo

rodneykinney committed:
Rename ExecuteShellCommand -> ExternalProcess

rodneykinney committed:
TrainModel example pipeline

rodneykinney committed:
Add python-based model-training example pipeline

rodneykinney committed:
Reorganize ExternalProcess

rodneykinney committed:
Rename Serializer/Deserializer.  Update README

rodneykinney committed:
Merge branch 'master' of https://github.com/allenai/pipeline

rodneykinney committed:
Rounded corners

rodneykinney committed:
Default AWS credentials

rodneykinney committed:
Merge branch 'master' of https://github.com/allenai/pipeline into artifact-factory

rodneykinney committed:
Merge branch 'master' of https://github.com/rodneykinney/pipeline into artifact-factory

Conflicts:
	src/main/scala/org/allenai/pipeline/ArtifactFactory.scala

rodneykinney committed:
Reformat

rodneykinney committed:
Merge branch 'master' into artifact-factory

Conflicts:
	src/main/scala/org/allenai/pipeline/ExternalProcess.scala

rodneykinney committed:
Refactor ArtifactFactory

rodneykinney committed:
Refactor ArtifactFactory

rodneykinney committed:
Rename, add comments

rodneykinney committed:
Clean up warnings

jefeweisen committed:
If ExternalProcess has OutputFileToken, check that it actually writes something.
Save newbies from this horrific call stack:

        6435     ERROR org.allenai.pipeline.Pipeline$$anon$1 [run-main-0] Untrapped exception
        java.io.FileNotFoundException: /var/folders/4j/7wz0c3gx7vgb340h0nf2cjsc0000gn/T/9174563377670596194/output (No such file or directory)
        	at java.io.FileInputStream.open(Native Method)
        	at java.io.FileInputStream.<init>(FileInputStream.java:138)
        	at org.allenai.pipeline.FileArtifact.read(FileArtifact.scala:22)
        	at org.allenai.pipeline.StreamIo$$anonfun$read$1.apply(ExternalProcess.scala:182)
        	at org.allenai.pipeline.StreamIo$$anonfun$read$1.apply(ExternalProcess.scala:182)
        	at org.allenai.pipeline.StreamIo$$anonfun$write$1.apply(ExternalProcess.scala:187)
        	at org.allenai.pipeline.StreamIo$$anonfun$write$1.apply(ExternalProcess.scala:185)
        	at org.allenai.pipeline.FileArtifact.write(FileArtifact.scala:31)
        	at org.allenai.pipeline.StreamIo$.write(ExternalProcess.scala:185)
        	at org.allenai.pipeline.StreamIo$.write(ExternalProcess.scala:180)
        	at org.allenai.pipeline.ProducerWithPersistence.create(Producer.scala:180)
        	at org.allenai.pipeline.Producer$$anonfun$7.apply(Producer.scala:64)
        	at org.allenai.common.Timing$.time(Timing.scala:19)
        	at org.allenai.pipeline.Producer$class.createAndTime(Producer.scala:64)
        	at org.allenai.pipeline.ProducerWithPersistence.createAndTime(Producer.scala:167)
        	at org.allenai.pipeline.Producer$class.org$allenai$pipeline$Producer$$cachedValue(Producer.scala:54)
        	at org.allenai.pipeline.ProducerWithPersistence.org$allenai$pipeline$Producer$$cachedValue$lzycompute(Producer.scala:167)
        	at org.allenai.pipeline.ProducerWithPersistence.org$allenai$pipeline$Producer$$cachedValue(Producer.scala:167)
        	at org.allenai.pipeline.Producer$class.get(Producer.scala:39)
        	at org.allenai.pipeline.ProducerWithPersistence.get(Producer.scala:167)
        	at org.allenai.pipeline.Pipeline$$anonfun$14.apply(Pipeline.scala:152)
        	at org.allenai.pipeline.Pipeline$$anonfun$14.apply(Pipeline.scala:152)
        	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
        	at scala.collection.immutable.List.foreach(List.scala:381)
        	at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
        	at scala.collection.immutable.List.map(List.scala:285)
        	at org.allenai.pipeline.Pipeline.runPipelineReturnResults(Pipeline.scala:152)
        	at org.allenai.pipeline.Pipeline.run(Pipeline.scala:25)

rodneykinney committed:
Create core/s3 sub-projects

rodneykinney committed:
Move S3 classes to s3 project

jefeweisen committed:
rename ExternalProcess.apply to ExternalProcess.a

jefeweisen committed:
comments

jefeweisen committed:
Move ExternalProcess.a to RunExternalProcess.a

jefeweisen committed:
Moves:

CommandOutput -> ExternalProcess.CommandOutput
ExtractOutputComponent -> RunExternalProcess.ExtractOutputComponent

jefeweisen committed:
Make default constructor of RunExternalProcess private to conceal knowledge of CommandOutput from consuming code.

jefeweisen committed:
Rename RunExternalProcess.a -> RunExternalProcess.apply

jefeweisen committed:
Fix oops: forgot to add these files

jefeweisen committed:
Implement versionHistory on RunExternalProcess

rodneykinney committed:
Add VersionedResource

rodneykinney committed:
Simplify artifact creation

rodneykinney committed:
PartitionedRddArtifact and Io

rodneykinney committed:
Rdd Persistence working in unit tests

rodneykinney committed:
Reformat

rodneykinney committed:
Cleanup

rodneykinney committed:
Rdd Object initialization and sample pipeline

rodneykinney committed:
Clean up style, logging, format

rodneykinney committed:
Add build.sbt

rodneykinney committed:
Update .gitignore

rodneykinney committed:
Mix-in traits to define urlToArtifacte

rodneykinney committed:
Clean up Pipeline factory methods

rodneykinney committed:
Merge pull request #2 from jefeweisen/check_for_external_process_output

If ExternalProcess has OutputFileToken, check that it actually writes…

rodneykinney committed:
Merge pull request #3 from jefeweisen/external_process_abstraction

External process abstraction

rodneykinney committed:
Merge pull request #4 from jefeweisen/external_process_versionHistory

External process version history

rodneykinney committed:
Merge branch 'master' of https://github.com/rodneykinney/pipeline into subprojects

rodneykinney committed:
Fix test

rodneykinney committed:
Make Pipeline a trait after all.

rodneykinney committed:
Resolve relative paths within ArtifactFactory

rodneykinney committed:
Add Pipeline.createOutputArtifact

rodneykinney committed:
Name steps in Pipeline.persist()

rodneykinney committed:
Bump version

rodneykinney committed:
autoGeneratedPath => hashId

rodneykinney committed:
Handle empty RDD on read

rodneykinney committed:
Fix tests

rodneykinney committed:
Remove type param from PartitionedRddArtifact

rodneykinney committed:
Fix CreateRddArtifacts

rodneykinney committed:
Fix step name resolution

rodneykinney committed:
Remove race condition

rodneykinney committed:
LineCollectionIo className

rodneykinney committed:
Disable caching for persisted RDDs

rodneykinney committed:
Deprecate Producer.persist()

rodneykinney committed:
Add runUntil. Custom wrapper for S3 credentials

rodneykinney committed:
Pipeline.persistCustom

rodneykinney committed:
Merge branch 'master' of https://github.com/allenai/pipeline

Conflicts:
	project/plugins.sbt
	version.sbt
